Mathlib/Topology/Homotopy/LocallyContractible.lean

```diff
@@ -0,0 +1,164 @@
+contractible neighborhoods to form a neighborhood basis. This is often called **strongly locally
+contractible** (SLC).
+
+**Hierarchy of notions:**
+* "Basis of open contractible neighborhoods" (strongest)
+* "Basis of contractible neighborhoods" (this file, SLC)
+* "Null-homotopic inclusions" (classical LC, weakest)
+
+The Borsuk-Mazurkiewicz counterexample [borsuk_mazurkiewicz1934] shows that classical LC does not
+imply SLC. Moreover, from a contractible neighborhood S one generally cannot shrink to an open
+V ⊆ S that remains contractible, so requiring neighborhoods to be open is potentially strictly
+stronger than SLC.
+-/
+
+noncomputable section
+
+open Topology Filter Set Function ContinuousMap
+
+variable {X Y : Type*} [TopologicalSpace X] [TopologicalSpace Y] {x y : X} {ι : Type*}
+
+section LocallyContractible
+
+/-- Classical **local contractibility**: for every point and every neighborhood U,
+there exists a neighborhood V ⊆ U such that the inclusion V ↪ U is null-homotopic.
+
+This is weaker than `StronglyLocallyContractibleSpace`. -/
+def LocallyContractible (X : Type*) [TopologicalSpace X] : Prop :=
+  ∀ (x : X) (U : Set X), U ∈ 𝓝 x →
+    ∃ (V : Set X) (hVU : V ⊆ U), V ∈ 𝓝 x ∧ Nullhomotopic (inclusion hVU)
+
+end LocallyContractible
+
+section StronglyLocallyContractibleSpace
+
+/-- A topological space is **strongly locally contractible** if, at every point, contractible
+neighborhoods form a neighborhood basis. Here "contractible" means contractible as a subspace.
+
+This is strictly stronger than the classical notion of locally contractible, which only requires
+null-homotopic inclusions. -/
+class StronglyLocallyContractibleSpace (X : Type*) [TopologicalSpace X] : Prop where
+  /-- Each neighborhood filter has a basis of contractible subspace neighborhoods. -/
+  contractible_basis : ∀ x : X,
+    (𝓝 x).HasBasis (fun s : Set X => s ∈ 𝓝 x ∧ ContractibleSpace s) id
+
+export StronglyLocallyContractibleSpace (contractible_basis)
+
+/-- A helper to construct a strongly locally contractible space from a neighborhood basis where
+each basis element is contractible as a subspace. -/
+theorem StronglyLocallyContractibleSpace.of_bases {p : X → ι → Prop} {s : X → ι → Set X}
+    (h : ∀ x, (𝓝 x).HasBasis (p x) (s x)) (h' : ∀ x i, p x i → ContractibleSpace (s x i)) :
+    StronglyLocallyContractibleSpace X where
+  contractible_basis x := by
+    rw [hasBasis_self]
+    intro t ht
+    obtain ⟨i, hpi, hi⟩ := (h x).mem_iff.mp ht
+    exact ⟨s x i, (h x).mem_of_mem hpi, h' x i hpi, hi⟩
+
+variable [StronglyLocallyContractibleSpace X]
+
+/-- In a strongly locally contractible space, for any open set U containing x, there is a basis of
+contractible neighborhoods of x contained in U. -/
+theorem contractible_subset_basis {U : Set X} (h : IsOpen U) (hx : x ∈ U) :
+    (𝓝 x).HasBasis (fun s : Set X => s ∈ 𝓝 x ∧ ContractibleSpace s ∧ s ⊆ U) id :=
+  (contractible_basis x).hasBasis_self_subset (IsOpen.mem_nhds h hx)
```

Comment:
It feels slightly strange to have this at all, since it's hardly shorter than inlining `(contractible_basis _).hasBasis_self_subset` below, and this version is provably equivalent to that. I won't require it to go away, as it's a least a statement a mathematician might expect to see.
```suggestion
/-- In a strongly locally contractible space, for any open set `U` containing `x`, there is a basis
of contractible neighborhoods of `x` contained in `U`. -/
theorem contractible_subset_basis {U : Set X} (h : IsOpen U) (hx : x ∈ U) :
    (𝓝 x).HasBasis (fun s : Set X => s ∈ 𝓝 x ∧ ContractibleSpace s ∧ s ⊆ U) id :=
  (contractible_basis x).hasBasis_self_subset (IsOpen.mem_nhds h hx)
```